File: solutions/euphoria/all-your-base/1/all-your-base.ex
include lib/errors.e

public function rebase(integer input_base, sequence input_digits, integer output_base)
  if input_base < 2 then
    return {ERR, "input base must be >= 2"}
  elsif output_base < 2 then
    return {ERR, "output base must be >= 2"}
  end if
  integer num = 0
  for i = 1 to length(input_digits) do
    if input_digits[i] < 0 or input_digits[i] >= input_base then
      return {ERR, "all digits must satisfy 0 <= d < input base"}
    end if
    num = num * input_base + input_digits[i]
  end for
  sequence res = {}
  while num > 0 do
    res = {remainder(num, output_base)} & res
    num = floor(num / output_base)
  end while
  if equal({}, res) then
    res = {0}
  end if
  return {OK, res}
end function

File: solutions/delphi/all-your-base/2/all_your_base.pas
unit uAllYourBase;

interface

uses
  Generics.Collections;

type
  TBase = class
    class function Rebase(aSourcebase: Integer; aSourceArray: TArray<Integer>; aTargetBase: Integer): TArray<Integer>;
  end;

implementation

uses
  SysUtils;

class function TBase.Rebase(aSourcebase: Integer; aSourceArray: TArray<Integer>; aTargetBase: Integer): TArray<Integer>;
var
  i, j, num: Integer;
  ResArray: TArray<Integer>;
begin
  // Validate bases
  if aSourcebase < 2 then
    raise EArgumentOutOfRangeException.Create('input base must be >= 2');
  if aTargetBase < 2 then
    raise EArgumentOutOfRangeException.Create('output base must be >= 2');

  // Convert sourcearray to a value
  num := 0;
  for i := Low(aSourceArray) to High(aSourceArray) do
  begin
    if (aSourceArray[i] >= 0) and (aSourceArray[i] < aSourcebase) then
      num := num * aSourcebase + aSourceArray[i]
    else
      raise EArgumentOutOfRangeException.Create('all digits must satisfy 0 <= d < input base');
  end;

  // Convert value to targetbase in array
  i := 0;
  repeat
    SetLength(ResArray, i + 1);
    ResArray[i] := num mod aTargetBase;
    num := num div aTargetBase;
    Inc(i);
  until num = 0;

  // Revert target array
  j := 0;
  SetLength(Result, Length(ResArray));
  for i := High(ResArray) downto 0 do
  begin
    Result[j] := ResArray[i];
    Inc(j);
  end;
end;

end.